onnxruntime/core/providers/cuda/tensor/transpose.cc

```diff
@@ -91,20 +91,22 @@ Status Transpose::DoTranspose(const Transpose& kernel,
   const std::vector<int64_t>& input_dims = input.Shape().GetDims();
   const std::vector<int64_t>& output_dims = output.Shape().GetDims();
 
-  auto rank = input_dims.size();
-  CudaAsyncBuffer<int64_t> input_strides(&kernel, rank);
-  CudaAsyncBuffer<size_t> perm(&kernel, permutations);
-  CudaAsyncBuffer<fast_divmod> fdm_output_strides(&kernel, rank);
-  ORT_ENFORCE(TensorPitches::Calculate(input_strides.CpuSpan(), input_dims));
-  ORT_ENFORCE(CalculateFdmStrides(fdm_output_strides.CpuSpan(), output_dims));
+  auto rank = static_cast<int32_t>(input_dims.size());
+  TensorPitches original_input_strides(input_dims);
+  TensorPitches original_output_strides(output_dims);
 
-  ORT_RETURN_IF_ERROR(input_strides.CopyToGpu());
-  ORT_RETURN_IF_ERROR(perm.CopyToGpu());
-  ORT_RETURN_IF_ERROR(fdm_output_strides.CopyToGpu());
+  TArray<int64_t> input_strides(rank);
+  for (auto i = 0; i < rank; i++) {
+    input_strides.data_[i] = original_input_strides[permutations[i]];
```

Comment:
Scott McKay
Would be better to put the ORT_ENFORCE in the TArray ctor instead of relying on everyone calling it to remember to do so.

Sherlock Huang
Yes.